ern-composite-gen/src/patchCompositeBabelRcRoots.ts

```diff
@@ -0,0 +1,84 @@
+import fs from 'fs-extra'
+import path from 'path'
+import semver from 'semver'
+import { log, readPackageJson, shell } from 'ern-core'
+import { getNodeModuleVersion } from './getNodeModuleVersion'
+import { patchMetroBabelRcRoots } from './patchMetroBabelRcRoots'
+
+export async function patchCompositeBabelRcRoots({ cwd }: { cwd: string }) {
+  const compositePackageJson = await readPackageJson(cwd)
+  const compositeNodeModulesPath = path.join(cwd, 'node_modules')
+  const compositeReactNativeVersion = await getNodeModuleVersion({
+    cwd,
+    name: 'react-native',
+  })
+  const compositeMetroVersion = await getNodeModuleVersion({
+    cwd,
+    name: 'metro',
+  })
+  const dependencies: string[] = Object.keys(compositePackageJson.dependencies)
+
+  // Any dependency that has the useBabelRc set in their package.json
+  // as follow ...
+  //
+  // "ern": {
+  //   "useBabelRc": true
+  // }
+  //
+  // ... is added to the babelRcRoots array, so that we can properly
+  // configure Babel to process the .babelrc of these dependencies.
+  const babelRcRootsRe: RegExp[] = []
+  const babelRcRootsPaths: string[] = []
+  for (const dependency of dependencies) {
+    if (await fs.pathExists(path.join(compositeNodeModulesPath, dependency))) {
+      const depPackageJson = await readPackageJson(
+        path.join(compositeNodeModulesPath, dependency)
+      )
+      if (depPackageJson.ern?.useBabelRc === true) {
+        babelRcRootsRe.push(new RegExp(`node_modules\/${dependency}(?!.+\/)`))
+        babelRcRootsPaths.push(`./node_modules/${dependency}`)
+      }
+    }
+  }
+
+  await patchMetroBabelRcRoots({
+    babelRcRootsRe,
+    cwd,
+    metroVersion: compositeMetroVersion,
+    rnVersion: compositeReactNativeVersion,
+  })
+
+  // If React Native version is less than 0.56 it is still using Babel 6.
+  // In that case .babelrc files will be processed in any node_modules
+  // which is not a desired behavior.
+  // Therefore we just remove all .babelrc from all node_modules.
+  // We only keep .babelrc of node_modules that are "whitelisted",
+  // as we want them to be processed.
```

Comment:
That's wild.. Did you run these tests locally? I assume it's such a specific scenario we don't have CI tests for it?

Reply:
Yes that is wild. This function was actually not introduced nor modified in any way part of this PR, it was just renamed. I don't think we have tests for this one, but it's been manually tested prior to initial release (more than a year back), and has been battle proofed since then (most of the MiniApps we have are having local .babelrc and thus all composites going from more than a year have been going through this code)

Reply:
Ooh okay got it. I saw it was just moved to a different file. :+1: